exchanges/bitstamp/bitstamp_websocket.go

```diff
@@ -3,126 +3,151 @@ package bitstamp
-		if enabledPair.String() == tradingPair {
-			return tradingPair, nil
-		}
-	}
-
-	return "", errors.New("bistamp_websocket.go error - could not find trading pair")
-}
-
 // WsConnect connects to a websocket feed
 func (b *Bitstamp) WsConnect() error {
 	if !b.Websocket.IsEnabled() || !b.IsEnabled() {
 		return errors.New(exchange.WebsocketNotEnabled)
 	}
 
-	tradingPairs = make(map[string]string)
-	var err error
-
+	var dialer websocket.Dialer
 	if b.Websocket.GetProxyAddress() != "" {
-		log.Warn("bitstamp_websocket.go warning - set proxy address error: proxy not supported")
+		proxy, err := url.Parse(b.Websocket.GetProxyAddress())
+		if err != nil {
+			return err
+		}
+		dialer.Proxy = http.ProxyURL(proxy)
 	}
 
-	b.WebsocketConn.Client, err = pusher.NewClient(BitstampPusherKey)
+	var err error
+	b.WebsocketConn, _, err = dialer.Dial(b.Websocket.GetWebsocketURL(), http.Header{})
 	if err != nil {
 		return fmt.Errorf("%s Unable to connect to Websocket. Error: %s",
-			b.GetName(),
+			b.Name,
 			err)
 	}
 
-	b.WebsocketConn.Data, err = b.WebsocketConn.Client.Bind("data")
-	if err != nil {
-		return fmt.Errorf("%s Websocket Bind error: %s", b.GetName(), err)
-
+	if b.Verbose {
+		log.Debugf("Successful connection to %v",
+			b.Websocket.GetWebsocketURL())
 	}
 
-	b.WebsocketConn.Trade, err = b.WebsocketConn.Client.Bind("trade")
+	err = b.seedOrderBook()
 	if err != nil {
-		return fmt.Errorf("%s Websocket Bind error: %s", b.GetName(), err)
+		b.Websocket.DataHandler <- err
 	}
-	b.GenerateDefaultSubscriptions()
-	go b.WsReadData()
 
-	for _, p := range b.GetEnabledCurrencies() {
-		orderbookSeed, err := b.GetOrderbook(p.String())
-		if err != nil {
-			return err
-		}
+	b.generateDefaultSubscriptions()
+	go b.WsReadData()
 
-		var newOrderBook orderbook.Base
+	return nil
+}
 
-		var asks []orderbook.Item
-		for _, ask := range orderbookSeed.Asks {
-			var item orderbook.Item
-			item.Amount = ask.Amount
-			item.Price = ask.Price
-			asks = append(asks, item)
+// WsReadData reads data coming from bitstamp websocket connection
+func (b *Bitstamp) WsReadData() {
+	b.Websocket.Wg.Add(1)
+	defer func() {
+		err := b.WebsocketConn.Close()
+		if err != nil {
+			b.Websocket.DataHandler <- fmt.Errorf("bitstamp_websocket.go - Unable to to close Websocket connection. Error: %s",
+				err)
 		}
+		b.Websocket.Wg.Done()
+	}()
 
-		var bids []orderbook.Item
-		for _, bid := range orderbookSeed.Bids {
-			var item orderbook.Item
-			item.Amount = bid.Amount
-			item.Price = bid.Price
-			bids = append(bids, item)
-		}
+	for {
+		select {
+		case <-b.Websocket.ShutdownC:
+			return
 
-		newOrderBook.Asks = asks
-		newOrderBook.Bids = bids
-		newOrderBook.Pair = p
-		newOrderBook.AssetType = "SPOT"
+		default:
+			b.Websocket.TrafficAlert <- struct{}{}
 
-		err = b.Websocket.Orderbook.LoadSnapshot(&newOrderBook, b.GetName(), false)
-		if err != nil {
-			return err
-		}
+			_, resp, err := b.WebsocketConn.ReadMessage()
+			if err != nil {
```

Comment:
Send err to DataHandler and break out of the for read loop as this is a permanent error